msgctxt support

gabegorelick opened the pull request:
Closes #4 

This is a **major** rewrite to how things are extracted, but **it's fully backwards compatible**. It achieves this by using the same technique as Jed. Namely, it stores messages with contexts as messages with their context prepended. Hopefully the code is self-explanatory, but I can provide more details if needed. This should also make it easier to add [domain support](https://github.com/smhg/gettext-handlebars/issues/2), although that will probably require a breaking change to the return value of `.parse`.

gabegorelick commented:
Went a little crazy and made a lot some general changes that aren't really related to this PR. Feel free to ignore them if you want. The important stuff is in [the first commit](https://github.com/gabegorelick/gettext-handlebars/commit/3873897070b7b2df9f1be24adb56c3ae62d1181f)

smhg commented:
Thanks for the PR. I've already cherry-picked most general changes.
I still have to take an in-depth look at the main changes (first commit).
They are making [gmarty/xgettext](https://github.com/gmarty/xgettext) fail. Is that something you would like to have a look at too?

gabegorelick commented:
Cool. I've rebased and updated this PR. I can look at why xgettext fails. It's probably due to some behavior it depends on that isn't verified in the unit tests, since all the tests still pass. 

Also, if we're modifying xgettext, are you OK with breaking backwards compatibility and bumping the major version? It's something that will need to be done to support domains anyway.

gabegorelick commented:
Fixed it. PR is updated and xgettext now works. Also added some more tests to verify the behavior xgettext expects.

smhg commented:
Thank you for all the work you put in to this. I still have to have a look at it.
Can you think of anything that would break because of this? Either in this project or xgettext(-template)?

gabegorelick commented:
Well all the existing tests still pass (for this project and
xgettext-template). It should be fine.

On Monday, April 6, 2015, Sam Hauglustaine <notifications@github.com> wrote:

> Thank you for all the work you put in to this. I still have to have a look
> at it.
> Can you think of anything that would break because of this? Either in this
> project or xgettext(-template)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/smhg/gettext-handlebars/pull/5#issuecomment-90198250>.
>

gabegorelick commented:
Any chance of getting this merged? I also want to add a lot more features (like domains, comment extraction, etc.), but I need this merged first.

smhg commented:
I'm sorry for the delay. I think it is better if I add you as a contributor (git + npm) so you can merge and publish this yourself.
Please also bump it to 0.4.0 because of the additional functionality.

joedrew commented:
Any chance for this to get merged? It's relatively important for correct translations.

smhg commented:
Hi @joedrew . These changes are kind of abandoned by @gabegorelick because he needed a different approach.
I wasn't able to check his PR. Would you be able to do that?
It seems to be complete at first sight. I'm happy to merge it if you could confirm they work for you. Please also test using [xgettext-template](https://github.com/gmarty/xgettext) since that's this module's main usage.

joedrew commented:
I've finally managed to get this PR to work with xgettext-template, but it took some changes to it to support non-digit keyword arguments:

```
diff --git a/bin/xgettext-template b/bin/xgettext-template
index 8929ed1..454a086 100644
--- a/bin/xgettext-template
+++ b/bin/xgettext-template
@@ -30,7 +30,7 @@ var argv = require('optimist')
   })
   .default({
     'from-code': 'ascii',
-    'keyword': ['_', 'gettext', 'ngettext:1,2'],
+    'keyword': ['_', 'gettext', 'ngettext:1,2', 'p_:msgctxt,msgid'],
     'force-po': false,
     'no-location': false
   })
diff --git a/index.js b/index.js
index 3b3ac05..91ce7f8 100644
--- a/index.js
+++ b/index.js
@@ -47,17 +47,14 @@ function xgettext(input, options, cb) {
 
         return parsers[name];
       },
-    specPattern = /([^:]+)(?::(\d)(?:,(\d))?)?/,
-    indexify = function (idx) {
-        return parseInt(idx, 10) - 1;
-      },
+    specPattern = /([^:]+)(?::([^,]+)(?:,([^,]+))?)?/,
     addToSpec = function (spec, keyword) {
         var parts = keyword.match(specPattern);
 
-        spec[parts[1]] = [parts[2] ? indexify(parts[2]) : 0];
+        spec[parts[1]] = [parts[2] ? parts[2] : 0];
 
         if (parts[3]) {
-          spec[parts[1]].push(indexify(parts[3]));
+          spec[parts[1]].push(parts[3]);
         }
 
         return spec;
```

Unfortunately, that further breaks xgettext-template. :(

smhg commented:
Thanks for your changes & tests!
Changes to xgettext-template are very much appreciated too. Even if it means things break.

smhg commented:
I think it is about time I catch up with this PR. I will try to have a look at things next week.

smhg commented:
All merged. New versions (also xgettext-template) available on npm. Please let me know if everything works as expected for you.
